dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/EventReferralValidationHookTest.java

```diff
@@ -0,0 +1,254 @@
+            .relationships( Arrays.asList( relationship ) )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 0, reporter.getReportList().size() );
+    }
+
+    @Test
+    public void verifyValidationSuccessIfReferralHasCompleteRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .relationships( Arrays.asList( relationship ) )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 0, reporter.getReportList().size() );
+    }
+
+    @Test
+    public void verifyValidationFailIfReferralHasInCompleteRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .relationships( Arrays.asList( relationshipIncomplete ) )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 1, reporter.getReportList().size() );
+        assertTrue( reporter.hasErrorReport( r -> r.getErrorCode() == TrackerErrorCode.E1312 ) );
+    }
+
+    @Test
+    public void verifyValidationFailIfReferralHasNoRelationShip()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) ) )
+            .thenReturn( programStageReferral );
+
+        Event event = Event.builder()
+            .event( CodeGenerator.generateUid() )
+            .program( MetadataIdentifier.ofUid( program ) )
+            .programStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_REFERRAL_UID ) )
+            .orgUnit( MetadataIdentifier.ofUid( ORG_UNIT_UID ) )
+            .attributeOptionCombo( MetadataIdentifier.EMPTY_UID )
+            .enrollment( ENROLLMENT_UID )
+            .build();
+
+        subject.validateEvent( reporter, bundle, event );
+
+        assertEquals( 1, reporter.getReportList().size() );
+        assertTrue( reporter.hasErrorReport( r -> r.getErrorCode() == TrackerErrorCode.E1311 ) );
```

Comment:
when checking size and content of a list I would use https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertAll-org.junit.jupiter.api.function.Executable...-

like

```suggestion
        assertAll( () -> assertEquals( 1, reporter.getReportList().size() ),
                   () -> assertTrue( reporter.hasErrorReport( r -> r.getErrorCode() == TrackerErrorCode.E1311 ) )
        );
```

so that in case the list contains 2 errors we still check whether it at least contains the error we want.

This gives us better error output. This way we don't have to re-run tests/debug them to figure out whats wrong. Super helpful also in CI when we have flaky tests and or with integration tests that just take so long to run.

Reply:
resolved